### ISSUE TITLE
Gradle is violating TestNG principle of test execution

### ISSUE THREAD

**Issue by prasanta-biswas** [gradle/gradle#17417]
Issue 1: Gradle is automatically calling the @BeforeSuite/@AfterSuite method defined in a class that is not in the test execution context. 
Example: 
Suppose Test.java and Base.java are  two classes. Base.java has defined @BeforeSuite and @AfterSuite method and Test.java extends Base.java. Now when I run test task from gradle, it is creating separate objects to call @BeforeSuite/@AfterSuite and @Test methods. This is violating TestNG principles as TestNG calls all these @Before/@After method using the same object which is used to call @Test method given the Base class is inherited by Test class.
However, when I'm running tests using TestNG directly, everything works as expected.

### Expected Behavior
If classes are related, then it should only create one Test object and call all annotated methods from the same object.

### Current Behavior
If the Base class is extended by the Test class, then Gradle should honour the inheritance and create only one child class object to call all annotated methods in the parent class.
### Context
I am trying to create a test framework using TestNG and Gradle. Defined multiple base classes which contain initialization code and test classes are extending those classes. Due to this issue, I'm getting a null pointer exception.

### Steps to Reproduce 
Tests class
```
package demo;

import org.testng.annotations.*;

public class Tests extends Base{
    
    @Test
    public void test(){
        System.out.println("Current Object: "+this.hashCode());
        System.out.println("TEST METHOD CALLED");
    }
}
```
Base Class
```
package demo;

import org.testng.annotations.AfterMethod;
import org.testng.annotations.AfterSuite;
import org.testng.annotations.BeforeMethod;
import org.testng.annotations.BeforeSuite;

public class Base{
    @BeforeSuite(alwaysRun = true)
    public void beforeSuite1(){
        System.out.println("Current Object: "+this.hashCode());
        System.out.println("BEFORE SUITE CALLED");
    }

    @AfterSuite(alwaysRun = true)
    public void afterSuite1(){
        System.out.println("Current Object: "+this.hashCode());
        System.out.println("AFTER SUITE CALLED");
    }
}
```
build.gradle
```
plugins {
    id 'java'
}

group 'org.example'
version '1.0-SNAPSHOT'

apply plugin: 'java'

repositories {
    mavenCentral()
}

task runTests(type: Test){
    useTestNG(){
        testLogging.exceptionFormat = 'full'
        testLogging.showStandardStreams = true
        useDefaultListeners = true
    }
}

dependencies {
    implementation "org.testng:testng:6.14.2"
}
```
When I run the above using Gradle task, I get the following output
```
Current Object: 2102746508
BEFORE SUITE CALLED
Current Object: 688041809
TEST METHOD CALLED
Current Object: 2102746508
AFTER SUITE CALLED
```
Look here the object IDs are different but according to TestNG, it should be the same object as the Test class is extending Base class.
When I run it using TestNG, I get the following
```
Current Object: 1879492184
BEFORE SUITE CALLED
Current Object: 1879492184
TEST METHOD CALLED
Current Object: 1879492184
AFTER SUITE CALLED
```
TestNG output is the correct one.
### Your Environment
Gradle version - 7.0.2
Java version - 1.8
IntellijIDEA version - 2020.1

**Comment by mlopatkin** [comment]
Could you please provide a reproducer as a self-contained archive or a github repo?

**Comment by prasanta-biswas** [comment]
@mlopatkin sure. I will put the sample project link shortly.

**Comment by prasanta-biswas** [comment]
@mlopatkin here is the sample project
https://github.com/prasanta-biswas/gradle-bug

**Comment by mlopatkin** [comment]
Thanks for the update!

**Comment by mlopatkin** [comment]
Unfortunately, I cannot reproduce the issue with the provided project:
```
> Task :runTests

Gradle suite STANDARD_OUT
    Current Object: 177649262
    BEFORE SUITE CALLED

Gradle suite > Gradle test > demo.gradle.TestClass2 STANDARD_OUT
    Current Object: 177649262
    BEFORE METHOD CALLED

Gradle suite > Gradle test > demo.gradle.TestClass2 > test STANDARD_OUT
    Current Object: 177649262
    TEST METHOD CALLED

Gradle suite > Gradle test > demo.gradle.TestClass2 STANDARD_OUT
    Current Object: 177649262
    AFTER METHOD CALLED

Gradle suite STANDARD_OUT
    Current Object: 177649262
    AFTER SUITE CALLED

BUILD SUCCESSFUL in 681ms
3 actionable tasks: 3 executed
```

I see that the same object is being used. Am I missing something?

**Comment by prasanta-biswas** [comment]
This is very strange. The same project I'm running. Getting a different output.

![](https://im7.ezgif.com/tmp/ezgif-7-29abf1ce0e84.gif)

**Comment by tonybasil21** [comment]
I tried this on my laptop and reproduced the same output as the one shared by @prasanta-biswas

**Comment by mlopatkin** [comment]
I was able to reproduce the issue with minor tweaks. What is interesting is that the Main class can exhibit the same behavior if I change the order of elements in the array:
```diff
diff --git a/src/test/java/demo/gradle/Main.java b/src/test/java/demo/gradle/Main.java
index f047f0a..5562016 100644
--- a/src/test/java/demo/gradle/Main.java
+++ b/src/test/java/demo/gradle/Main.java
@@ -5,7 +5,7 @@ import org.testng.TestNG;
 public class Main {
     public static void main(String[] args) {
         TestNG runner = new TestNG();
-        runner.setTestClasses(new Class[]{TestClass2.class, TestClass1.class});
+        runner.setTestClasses(new Class[]{TestClass1.class, TestClass2.class});
         runner.run();
     }
 }
```

Then the output becomes
```
> Task :Main.main()
Current Object: 510113906
BEFORE SUITE CALLED
Current Object: 2008362258
BEFORE METHOD CALLED
Current Object: 2008362258
TEST METHOD CALLED
Current Object: 2008362258
AFTER METHOD CALLED
Current Object: 510113906
AFTER SUITE CALLED
```
I assume that the behavior of the Gradle `runTests` task depends on the order in which classes are added to the list.

I'm by no means a TestNG expert but the behavior of the changed Main class suggests that your assumption about `BeforeSuite` being executed for the same instance as the test may be incorrect. The [answer from the TestNG maintainer](https://stackoverflow.com/a/3411597/15894854) also supports this point with "`@BeforeSuite` is only run once, regardless of what class it belongs to", albeit weakly. Could you please point me out the docs stating that BeforeSuite should be executed for the same instance?

**Comment by prasanta-biswas** [comment]
@mlopatkin thanks for the link to the Stackoverflow answer. I think I have mixed it up with inheritance. I was under the assumption that it should maintain the hierarchy for calling the methods with the same test instance. I was trying with TestNG runner plugin in IntelliJIDEA and it was working as I mentioned and Gradle runner gave me NullPointerException for the same code as I was initializing an object in @BeforeSuite and was using that in @BeforeMethod. 
Strangely TestNG runner worked and Gradle failed.

Thanks a lot for your time. Closing the issue.
